bluetooth: fast_pair: Add bond management

alstrzebonski opened the pull request:
Implements Fast Pair bond management. Associates bonds with Account Keys
and removes bonds when associated Account Keys are removed.

Short solution description: https://nordicsemi.atlassian.net/wiki/spaces/SPG/pages/269811994/New+Fast+Pair+storage+-+chosen+design#1.-Backwards-compatibility---standard-implementation%3A (see point 1.3. Bond information)

Detailed bond management flowchart: https://nordicsemi.atlassian.net/wiki/spaces/SPG/pages/672793094
![bond_flow drawio (1) (1)](https://github.com/nrfconnect/sdk-nrf/assets/86837740/c166caaa-a45a-4496-8e0f-5031c5c9174c)


Jira: NCSDK-18829

TODO in scope of this PR:
- [x] Align Fast Pair storage tests (currently there are build failures)
- [x] Introduce Kconfig option that will enable/disable bond tracking functionality (should also fix tests build failures)
- [x] Perform detailed tests of all edge cases
- [x] Improve function documentation in header
TODO in separate PR:
- [ ] Add documentation.

NordicBuilder commented:
## Test specification

### [CI/Jenkins/NRF](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-15146/1/)
- Integration Platforms

### [CI/Jenkins/integration](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-15146/1/)
| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| desktop52_verification | X | | |
| test-fw-nrfconnect-ble | X | | |
| test-fw-nrfconnect-ble_samples | X | | |
| test-sdk-find-my | X | | |


[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sdk-nrf/job/PR-15146/1/artifact/test-spec/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak.h:
```diff
@@ -89,6 +93,24 @@ int fp_storage_ak_get(struct fp_account_key *buf, size_t *key_count);
 int fp_storage_ak_find(struct fp_account_key *account_key,
 		       fp_storage_ak_check_cb account_key_check_cb, void *context);
 
+/** Save potential bond.
```
You can also carry the doc improvements for this API docs

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
Could we separate this new bond management logic into a dedicated source file e.g. fp_storage_ak_bond.c?

In this scenario, the FP storage AK could depend on this module and interact with it in this file. The fp_storage_ak_bond.c unit would be a helper storage unit of the fp_storage_ak.c.

I see a few places where interaction between fp_storage_ak.c and the fp_storage_ak_bond.c is needed:

- `ak_metadata_id_get` getting Account Key ID encoding into bond NVM data. You can extract this information and pass the ID to the new helper module API.
-  passing the information about the Account Key overwrite. You can request the bond deletion.

MarekPieta replied:
I would suggest to keep the functionalities together. `fp_storage_ak` is tightly coupled with the new functionality. Moving part of it to another module would make NVM data integrity checks more difficult.

MarekPieta replied:
Discussed offline. Not implementing for now.

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
To avoid dependency on the Bluetooth subsystem, you can move the Bluetooth management logic to the upper layer and define the notification API that the upper layer (fp_bond_management.c) is required to use:

```
fp_storage_ak_bond_bt_disconnected_notify(uint8_t id, uint8_t *addr);

fp_storage_ak_bond_bt_identity_resolved_notify(uint8_t id, uint8_t *old_addr, uint8_t *new_addr);

struct fp_storage_ak_bond_bt_request_cb {
	int (*unpair_request)(uint8_t id, uint8_t *addr);

	int (*bond_find)(uint8_t id, uint8_t *addr)
}

fp_storage_ak_bond_bt_request_cb_register(struct fp_storage_ak_bond_bt_request_cb *cb);
```

You may even be able to drop the Bluetooth Identity argument from most of this function as the upper layer is probably able to deduce it from the address

MarekPieta replied:
I would also opt for removing direct dependencies on Bluetooth stack from `fp_storage`. Still, I would suggest simpler API:

- update bond info (address, confirmed or not, account key ID)
- update address (old address, new address)

A simple confirm mechanism could be good enough for this use-case. The only thing that needs to be done is to notify Fast Pair modules about Account Key removal and about invalid bond that is dropped from FP storage.

alstrzebonski replied:
```
fp_storage_ak_bond_bt_disconnected_notify(uint8_t id, uint8_t *addr);

fp_storage_ak_bond_bt_identity_resolved_notify(uint8_t id, uint8_t *old_addr, uint8_t *new_addr);

struct fp_storage_ak_bond_bt_request_cb {
	int (*unpair_request)(uint8_t id, uint8_t *addr);

	int (*bond_find)(uint8_t id, uint8_t *addr)
}

fp_storage_ak_bond_bt_request_cb_register(struct fp_storage_ak_bond_bt_request_cb *cb);
```

kapi-no reviewed (commented):
Just finished a high-level overview of the code. I suggested some improvements to the issues that came up during our previous offline discussions:

- Separate the management of the AK bond association data from the fp_storage_ak.c to the dedicated helper module. It could cleanly separate the code for storing AK data and the AK bond association data. The separation should allow us to easily introduce Kconfig for activating this feature.
- Move the Bluetooth subsys dependency outside the storage unit to the upper layer (e.g. fp_bond_manager.c). This change should make it easier for us to test the storage unit for AK bond associate data.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -21,12 +23,27 @@
+/* Non-volatile bond information that is stored in the Settings subsystem. */
+struct fp_bond_info_nv {
+	uint8_t account_key_metadata_id;
+	uint8_t identity;
```
I wonder if we actually need to store identity here. Isn't it already done by Bluetooth stack?

alstrzebonski replied:
It's needed when calling bt_unpair. Otherwise you would have to search through all identities with `bt_foreach_bond` to find the identity of the bond. Considering it's only 1 byte, we've decided it's easier to store it here.

alstrzebonski replied:
After the discussion we eventually decided not to store identity here.

alstrzebonski commented on tests/subsys/bluetooth/fast_pair/storage/account_key_storage/src/settings_mock.c:
```diff
@@ -11,6 +11,9 @@
 #include "storage_mock.h"
 
+/* Ensure 'strnlen' is available even with -std=c99. */
+size_t strnlen(const char *buf, size_t bufsz);
```
It replicates on my machine when building normally with west build

MarekPieta replied:
For which board? What SDK do you use?

kapi-no replied:
Zephyr SDK v0.16.6 solves this issue

alstrzebonski replied:
Yes, updating the SDK solved the issue. I'm dropping this commit.

alstrzebonski commented:
Rebased on main.

alstrzebonski commented:
Addressed review comments apart from API description - it will be added once we agree on the API. Please rereview. Focus on general structure, API, Kconfigs and interactions between modules (including Bluetooth subsys).

alstrzebonski commented:
Rebased on main.

alstrzebonski commented:
Added small fixes.

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -473,6 +1037,14 @@ static int fp_storage_ak_reset(void)
+	const struct fp_storage_ak_bond_bt_request_cb *registered_bt_request_cb = bt_request_cb;
+
+	if (was_enabled) {
+		err = fp_storage_ak_uninit();
```
You can consider merging the independent improvements in a separate PR (like this one or fp_auth.h doc improvement) that could be potentially merged within one day.

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +714,248 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+		}
+	} else {
+		/* The actual ID will be determined during Account Key write. */
+		id = 0;
```
Consider creating a define to distinguish this zero value (e.g AK_ID_UNPOPULATED)

MarekPieta replied:
Discussed offline, we can use `AK_ID_UNKNOWN` here

alstrzebonski replied:
Maybe `BOND_AK_ID_UNKNOWN` to point that it is used only in bond entries.

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -383,6 +656,35 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+
+			bond.nv.account_key_metadata_id = id;
+
+			err = bond_name_gen(bond_name, bond_idx);
+			if (err) {
+				return err;
+			}
+
+			err = settings_save_one(bond_name, &bond.nv, sizeof(bond.nv));
+			if (err) {
+				return err;
+			}
+
+			fp_bonds[bond_idx] = bond;
```
You use this call sequence at least 3 times. Could you define a helper function to avoid duplication?

MarekPieta replied:
If we opt for relying on `struct fp_bond *` instead of index, we could simplify internal APISs further. We could make a rollback copy and then update  record in an array directly through a pointer (on NVM error we would assign back the rollback copy). Storing function could then take only `struct fp_bond *` as an argument (index could be calculated internally as pointer diff; the function might even perform delete under the hood is data was zeroed in RAM).

Initially discussed the idea with @alstrzebonski and he was sceptical about it (mentioning inconsistency with other storage modules and introducing unnecessary changes here). Still, seemingly only this module does read->update->store, so using rollback mechanism here is justified in my opinion (other modules overwrite previous record with new data).

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +296,186 @@ static int validate_ak_order(void)
+
+	handle_bond_duplicates();
+
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
```
Perhaps also define a helper function for this code block. e.g `invalid_bonds_purge()`

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +296,186 @@ static int validate_ak_order(void)
+
+	bt_bonds_find();
+
+	handle_bond_duplicates();
```
```suggestion
	bond_duplicates_handle();
```

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/Kconfig.fp_storage:
```diff
@@ -34,6 +34,15 @@ config BT_FAST_PAIR_STORAGE_OWNER_ACCOUNT_KEY
 
+config BT_FAST_PAIR_STORAGE_BOND_MANAGEMENT
+	bool
+	depends on BT
```
I think we can remove the dependency here. The Bluetooth address header does not count as it is always included regardless of the `BT` Kconfig option value

alstrzebonski replied:
We have following dependency:
```
#if CONFIG_BT_FAST_PAIR_STORAGE_BOND_MANAGEMENT
#define FP_BONDS_ARRAY_LEN	(CONFIG_BT_MAX_PAIRED + CONFIG_BT_MAX_CONN)
#else
#define FP_BONDS_ARRAY_LEN	0
#endif
```
If we enable the `CONFIG_BT_FAST_PAIR_STORAGE_BOND_MANAGEMENT` without `BT`, it will lead to build failure.

MarekPieta replied:
Are the build failures related to these Kconfigs or to some other missing C definitions?

kapi-no reviewed (commented):
Generally looks good, I will follow up with a more detailed review once the PR is taken out of the draft stage and other reviewers approve the high-level architecture of this module.

alstrzebonski commented on subsys/bluetooth/services/fast_pair/include/fp_keys.h:
```diff
@@ -179,6 +179,16 @@ void fp_keys_bt_auth_progress(const struct bt_conn *conn, bool authenticated);
  */
 void fp_keys_drop_key(const struct bt_conn *conn);
 
+/** Save potential bond in the Account Key storage module.
+ *  The Keys module adds the Account Key information to the bond.
+ *  The bond will be deleted if the pairing fails or the Fast Pair Procedure is incomplete.
+ *
+ * @param[in] conn Potential bond's connection object.
+ *
+ * @return 0 If the operation was successful. Otherwise, a (negative) error code is returned.
+ */
```
Yes, it will be fixed when we agree on API.

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/include_priv/fp_storage_ak_priv.h:
```diff
@@ -46,6 +46,24 @@ extern "C" {
 	(sizeof(SETTINGS_AK_FULL_PREFIX) + SETTINGS_AK_NAME_MAX_SUFFIX_LEN)
 
+/** Settings key name prefix for bond. */
+#define SETTINGS_BOND_NAME_PREFIX "bond"
```
I can rename it to `bond_info`.

MarekPieta commented on subsys/bluetooth/services/fast_pair/Kconfig.fast_pair:
```diff
@@ -172,6 +172,29 @@ config BT_FAST_PAIR_STORAGE_INTEGRATION_INIT_PRIORITY
+config BT_FAST_PAIR_BOND_MANAGER
+	bool "Fast Pair bond management functionality"
+	default y
+	depends on !BT_FAST_PAIR_FMDN
```
Or maybe a storage backend could expose information if it supports the feature (something like `CONFIG_BT_FAST_PAIR_STORAGE_OWNER_ACCOUNT_KEY` for Owner AK support)

alstrzebonski replied:
It causes dependency loop. 
![image](https://github.com/user-attachments/assets/cb3696ec-33c4-443b-825f-511739d57bce)
I'd keep it as is.

alstrzebonski replied:
We will remove the dependency once we add the support for it.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_bond_manager.c:
```diff
@@ -0,0 +1,120 @@
+
+static int bond_manager_init(void)
+{
+	fp_storage_ak_bond_bt_request_cb_register(&bt_request_cb);
```
Since `bond_find` callback is called only during intialization, I would suggest to make it an API function. - e.g. `fp_storate_ak_bond_load_ble_bond(address)`.

MarekPieta replied:
Then we would need to invert the initialization priorities and delay calling `bond_remove` callbacks until the `fp_storage_ak_bond_bt_request_cb_register` is called

MarekPieta replied:
Discussed offline with @alstrzebonski and @kapi-no. We will stick with the existing approach.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +714,248 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+		id = 0;
+	}
+
+	err = bond_free_idx_get(&bond_idx);
```
Maybe you could return pointer to found `struct fp_bond` and fill it directly to simplify implementation and avoid copying? I think you could also safely assert on error here

MarekPieta replied:
You could event make something like `fp_bond_find(conn_ctx)` where `conn_ctx == NULL` would be used to find free structure

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +714,248 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+	fp_bonds[bond_idx].bonded = true;
+
+	while (!duplicate_find_err) {
+		duplicate_find_err = bonded_duplicate_find(&bond_duplicate_idx, addr, conn_ctx);
```
We may allow to perform this check only once and then assert that no duplicate is found I think (current implementation may get stuck in a never ending loop...)

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +714,248 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+		return;
+	}
+
+	bond_update_addr(new_addr, conn_ctx);
```
We could directly write to field representing address of proper conn and trigger NVM store here

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -481,13 +1053,37 @@ static int fp_storage_ak_reset(void)
+
+			err = bt_request_cb->bond_remove(&fp_bonds[index].nv.addr);
+			if (err) {
+				LOG_WRN("Failed to unpair (err %d). This might happen in some edge "
```
In case something went really wrong here we should not remove bond's AK association I think (BLE bond would still be here)

alstrzebonski replied:
FYI @MarekPieta: The edge case may happen if the `fp_storage_factory_reset` is interrupted by power down after calling `bt_unpair` and before getting `bond_deleted` callback. Then, when whole storage is being enabled, the storage reset is resumed without going through regular storage validation and we may try to call `bt_unpair` on already unpaired bond. I will separate those log messages and explain the edge case in code comment.

MarekPieta replied:
Thanks for clarification. Can we detect this case relying on the returned value? If so, I think it may be worth to handle it separately.

alstrzebonski replied:
bt_unpair doesn't return error when someone tries to unpair already unpaired bond. I will levae the comment in code.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +296,186 @@ static int validate_ak_order(void)
+	return 0;
+}
+
+static int bond_delete(size_t bond_idx)
```
`ble_bond_nvm_delete`

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +296,186 @@ static int validate_ak_order(void)
+	return 0;
+}
+
+static int bond_remove(size_t bond_idx)
```
`ble_bond_stack_unpair` ?

alstrzebonski replied:
Maybe `bond_remove_completely`. This function not only unpairs the bond but also deletes the FP bond from settings if it's not bonded.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_bond_manager.c:
```diff
@@ -0,0 +1,120 @@
+
+BT_CONN_CB_DEFINE(conn_callbacks) = {
+	.disconnected = disconnected,
+	.identity_resolved = identity_resolved,
```
As an alternative, we can document requirement that Fast Pair must be kept enabled when performing Bluetooth operations (we cannot enforce that in implementation).

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_bond_manager.c:
```diff
@@ -0,0 +1,120 @@
+{
+	struct bond_find_ctx *ctx = user_data;
+
+	if (ctx->bond_found) {
```
Discussed offline. We will not store local identity - assuming that we can have only one BLE bond with given peer.

MarekPieta commented on subsys/bluetooth/services/fast_pair/Kconfig.fast_pair:
```diff
@@ -172,6 +172,29 @@ config BT_FAST_PAIR_STORAGE_INTEGRATION_INIT_PRIORITY
+	bool "Fast Pair bond management functionality"
+	default y
+	depends on !BT_FAST_PAIR_FMDN
+	select BT_FAST_PAIR_STORAGE_BOND_MANAGEMENT
```
`BT_FAST_PAIR_STORAGE_AK_BOND_MANAGEMENT`

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak.h:
```diff
@@ -102,6 +106,39 @@ int fp_storage_ak_find(struct fp_account_key *account_key,
+
+void fp_storage_ak_bond_stopped(const void *conn_ctx);
+
+void fp_storage_ak_bond_update(const void *conn_ctx, const bt_addr_le_t *new_addr);
```
`fp_storage_ak_bond_updated`

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak.h:
```diff
@@ -102,6 +106,39 @@ int fp_storage_ak_find(struct fp_account_key *account_key,
+
+void fp_storage_ak_bond_bt_request_cb_register(const struct fp_storage_ak_bond_bt_request_cb *cb);
+
+void fp_storage_ak_bond_stopped(const void *conn_ctx);
```
`fp_storage_ak_bond_procedure_aborted`

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak.h:
```diff
@@ -102,6 +106,39 @@ int fp_storage_ak_find(struct fp_account_key *account_key,
+ *
+ * @return 0 If the operation was successful. Otherwise, a (negative) error code is returned.
+ */
+int fp_storage_ak_bond_save(const bt_addr_le_t *addr, const void *conn_ctx,
```
`fp_storage_ak_bond_create`

MarekPieta replied:
`bond` -> `bond_conn` for APIs used by "temporary bond"

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak.h:
```diff
@@ -102,6 +106,39 @@ int fp_storage_ak_find(struct fp_account_key *account_key,
+ *
+ * @param[in] conn Bond's connection object.
+ */
+void fp_storage_ak_bond_confirm(const bt_addr_le_t *addr, const void *conn_ctx);
```
`fp_storage_ak_bond_confirmed`

MarekPieta replied:
Also make sure to pass conn context as the first argument in these functions

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak.h:
```diff
@@ -102,6 +106,39 @@ int fp_storage_ak_find(struct fp_account_key *account_key,
+
+void fp_storage_ak_bond_update(const void *conn_ctx, const bt_addr_le_t *new_addr);
+
+void fp_storage_ak_bond_deleted(const bt_addr_le_t *addr);
```
`fp_storage_ak_bond_delete`

alstrzebonski commented:
Rebased on main to include https://github.com/nrfconnect/sdk-nrf/pull/16554

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +296,186 @@ static int validate_ak_order(void)
+		}
+	}
+
+	bond_remove_old_id();
```
`bonds_with_invalid_id_remove_completely`? We doesn't remove bonds with AK set to 0 here.

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -383,6 +656,35 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+	if (conn_ctx) {
+		__ASSERT_NO_MSG(IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_BOND_MANAGEMENT));
+
+		err = bond_find(&bond_idx, conn_ctx);
```
As for `ak_metadata_id_get` i would leave it as is to avoid casting.

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +296,186 @@ static int validate_ak_order(void)
+static void bt_bonds_find(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		if (!bt_request_cb->bond_find(&fp_bonds[i].nv.addr)) {
```
maybe `bool (*is_addr_bonded)(const bt_addr_le_t *addr);`?

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +714,248 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+		id = 0;
+	}
+
+	err = bond_free_idx_get(&bond_idx);
```
I reworked it to `static struct fp_bond *fp_bond_free_get(void)`

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +714,248 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+{
+	__ASSERT_NO_MSG(IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_BOND_MANAGEMENT));
+
+	if (!is_enabled) {
```
It will be handled on enable operation. We check whether the bond is bonded in Bluetooth stack and take appropriate actions.

alstrzebonski replied:
But generally the disable shouldn't happen before the procedure is finished.

alstrzebonski commented:
Addressed review comments in following push: https://github.com/nrfconnect/sdk-nrf/compare/fee6b45c958dfaa7e38507450ce84d7b585d7072..357a4b7800af8027229e0899ee3717ecbf3c138a). Please rereview.

alstrzebonski commented:
Fixed compliance issues.

alstrzebonski commented:
Added small kconfig improvement.

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/include_priv/fp_storage_ak_priv.h:
```diff
@@ -46,6 +46,30 @@ extern "C" {
 	(sizeof(SETTINGS_AK_FULL_PREFIX) + SETTINGS_AK_NAME_MAX_SUFFIX_LEN)
 
+/** Settings key name prefix for bond info. */
+#define SETTINGS_BOND_NAME_PREFIX "bond_info"
```
Consider naming the defines as BOND_INFO for consistency

MarekPieta replied:
I wonder if it should actually be `bond_info` or just `bond` (the `bond` term could refer to the data stored in NVM, `bond_info` is the NVM data + RAM data that is used during store handling)

alstrzebonski replied:
After the rename the struct is also named bond_info

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak_bond.h:
```diff
@@ -0,0 +1,142 @@
+ *  Used only when the CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND Kconfig option is enabled.
+ *
+ *  It confirms the potential Bluetooth bond associated with the Account Key. The association should
+ *  be earlier created using the @ref fp_storage_ak_bond_conn_create.
```
```suggestion
 *  be earlier created using the @ref fp_storage_ak_bond_conn_create API.
```

Also in other places

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -393,6 +715,14 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+
+	if (IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND) &&
+	    bond->bonded) {
+		/* Procedure finished successfully. Setting conn_ctx to NULL. */
+		bond->conn_ctx = NULL;
```
Shouldn't we remove the bond in case it is stored with the RPA address instead of the Identity Address?

alstrzebonski replied:
I think it's not possible to get pairing_complete BT callback with the RPA address

MarekPieta replied:
Should not be possible. Since we store it in NVM it may be worth to assert that assumption.

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -473,6 +1060,7 @@ static int fp_storage_ak_reset(void)
 {
 	int err;
 	bool was_enabled = is_enabled;
+	const struct fp_storage_ak_bond_bt_request_cb *registered_bt_request_cb = bt_request_cb;
```
I think we shouldn't reset the callback during factory reset to preserve the behaviour from other Fast Pair modules. We can consider them as a special RAM object that is not affected by factory reset. BTW, if you get an error in the `fp_bond_settings_delete`, you will lose the callback context.

alstrzebonski replied:
We reinstate it after ram_clear. Here we save it for later to call fp_storage_ak_bond_bt_request_cb_register

alstrzebonski replied:
I should call `fp_storage_ak_bond_bt_request_cb_register(registered_bt_request_cb);` after `fp_storage_ak_ram_clear` regardless of the `was_enabled` variable. Then it will be fine.

kapi-no replied:
discussed offline

```
if (IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND)) {
	registered_bt_request_cb = bt_request_cb;
}

fp_storage_ak_ram_clear();

if (IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND)) {
	bt_request_cb = registered_bt_request_cb;
}
```

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -488,13 +1076,45 @@ static int fp_storage_ak_reset(void)
 	}
 
+	if (IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND)) {
+		for (uint8_t index = 0; index < ARRAY_SIZE(fp_bonds); index++) {
```
you consider defining an iterator instead of a standard for loop to make the code more modular

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +735,250 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+	}
+}
+
+static struct fp_bond *bonded_by_addr_get(const bt_addr_le_t *addr)
```
Group the getter functions together and name to indicate the similarity

fp_bond_info_by_addr_get(...
fp_bond_info_by_conn_get(...
fp_bond_info_duplicate_get(...)
fp_bond_info_free_get(...)

MarekPieta replied:
Or maybe even `xyz_to_fp_bond` for the conversions?

alstrzebonski replied:
I'll stick with Kamil's approach

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +277,211 @@ static int validate_ak_order(void)
+	}
+}
+
+static void bt_bonds_find(void)
```
```suggestion
static void fp_bond_info_bonded_field_set_foreach(void)
```

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +277,211 @@ static int validate_ak_order(void)
+	}
+}
+
+static void bonds_conn_ctx_reset(void)
```
```suggestion
static void bonds_conn_ctx_reset(void)
```
```suggestion
static void fp_bond_info_conn_ctx_field_reset_foreach(void)
```

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +277,211 @@ static int validate_ak_order(void)
+			continue;
+		}
+
+		if (fp_bonds[i].nv.account_key_metadata_id == BOND_AK_ID_UNKNOWN) {
+			/* Fast Pair Procedure unfinished. */
+			err = bond_remove_completely(&fp_bonds[i]);
+			if (err) {
+				LOG_ERR("Failed to remove bond (err %d)", err);
+			}
+		}
```
Isn't this case already covered by the bonds_with_invalid_id_remove_completely function?

alstrzebonski replied:
No, it only removes ids that are associated with old, unexistant AKs. It doesn't remove BOND_AK_ID_UNKNOWN ids. The `bonds_with_invalid_id_remove_completely` function is also used in `fp_storage_ak_save` where we don't want to remove bond entries with equal to BOND_AK_ID_UNKNOWN.

kapi-no replied:
ok

MarekPieta replied:
Can we combine this function and `bonds_with_invalid_id_remove_completely` ?

kapi-no replied:
@MarekPieta, please resolve this comment during your review as you suggested an improvement in its scope

kapi-no reviewed (commented):
Looking good

kapi-no commented:
Resolved my old comments

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_auth.c:
```diff
@@ -235,6 +258,15 @@ static void pairing_failed(struct bt_conn *conn, enum bt_security_err reason)
 	clear_conn_context(conn);
 }
 
+static void bond_deleted(uint8_t id, const bt_addr_le_t *peer)
```
Yes, my bad

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak_bond.h:
```diff
@@ -0,0 +1,142 @@
+ *
+ *  The module links the Account Key information with the Bluetooth bond and stores
+ *  it in the NVM memory. The association and the Bluetooth bond will be deleted if the bond is not
+ *  confirmed using the @ref fp_storage_ak_bond_conn_confirm API before the bonding process is
+ *  cancelled using the @ref fp_storage_ak_bond_conn_cancel API.
```
That's not true. We call `fp_storage_ak_bond_conn_cancel` on disconnection. Then we check if the bond is confirmed. If yes - do nothing. Otherwise delete it.

alstrzebonski replied:
the word "cancel" here is a bit controversial but we decided it's better than "stopped", because it clearly indicates that it won't be resumed.

kapi-no replied:
`conn_finalize` as this API is used in the success case and the failure case. CC: @MarekPieta

MarekPieta replied:
`finalize` may not be the most fortunate word too (it's actually not needed to finish the store operation). Maybe we should simply avoid calling it once we confirmed bond in `fp_keys`? That may make code more intuitive.

alstrzebonski replied:
I'd let the storage module decide what to do with the bond after the conn is disconnected. Let's call it `finalize` if nothing betters comes to mind.

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +277,211 @@ static int validate_ak_order(void)
+	return 0;
+}
+
+static void bonds_with_invalid_id_remove_completely(void)
```
It inherits it from `bond_remove_completely` function

MarekPieta replied:
The `completely` suffix seems unnecessary. I will try to propose some renames to make APIs more straightforward here.

MarekPieta replied:
Maybe `bonds_with_invalid_ak_id_delete`

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +277,211 @@ static int validate_ak_order(void)
+	return 0;
+}
+
+static void bonds_with_invalid_id_remove_completely(void)
```
I don't want to fail `bt_fast_pair_init` because of that

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +735,250 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+
+	bond->bonded = true;
+
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		bond_duplicate = bonded_duplicate_get(conn_ctx, addr);
+		if (bond_duplicate) {
+			err = fp_bond_delete(bond_duplicate);
+			if (err) {
+				LOG_ERR("Failed to delete bond (err %d)", err);
+				break;
+			}
+
+			LOG_DBG("Deleted duplicated bond");
+		} else {
+			break;
+		}
+	}
```
this solution does not provide a way to break the iteration process (needs additional param)

kapi-no replied:
The convenience macro for the iteration process is not really what I meant here but it is a step in the right direction. Resolving as this is "consider" type comment

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +277,211 @@ static int validate_ak_order(void)
+	return bond - fp_bonds;
+}
+
+static int fp_bond_settings_delete(struct fp_bond *bond)
```
Function naming is getting a bit counterintuitive here. Maybe we can create one function instead of `fp_bond_settings_delete`, `fp_bond_delete` and `bond_remove_completely`. Separating `fp_bond_settings_delete` may not be necessary. The new function could take `bool ble_stack_bond_remove` as an argument to decide if we need to remove bonding information from BLE stack).

alstrzebonski replied:
Maybe `fp_bond_info_settings_delete(struct fp_bond_info *bond)` and `bond_remove(struct fp_bond_info *bond, bool bt_stack_bond_remove)`?

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +277,211 @@ static int validate_ak_order(void)
+{
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		fp_bonds[i].bonded = bt_request_cb->is_addr_bonded(&fp_bonds[i].nv.addr);
+		if (fp_bonds[i].bonded) {
+			LOG_DBG("Loaded bond has been found in the Bluetooth bond list");
+		}
```
The log is not too informative. Consider either using `LOG_WRN` to indicate case were bond was not found or logging BLE address here (you can also drop the log completely).

alstrzebonski replied:
It helps in debugging.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +277,211 @@ static int validate_ak_order(void)
+	 * Procedure disconnected, there may be a conn_ctx not set to NULL. All conn_ctx should be
+	 * set to NULL at this point.
+	 */
+	bonds_conn_ctx_reset();
```
This does not seem like part of bond validation. Can we do it explicitly on init/uninit?

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +277,211 @@ static int validate_ak_order(void)
+	 */
+	bonds_conn_ctx_reset();
+
+	bt_bonds_find();
```
Does not seem like part of actual bond validation logic. Can we do it explicitly on init?

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,6 +277,211 @@ static int validate_ak_order(void)
+	bonds_with_invalid_id_remove_completely();
+}
+
 static int fp_settings_validate(void)
```
Consider making functions like `fp_settings_validate_ak` and `fp_settings_validate_fp_bond` (to maintain the same level of abstraction)

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -349,7 +611,41 @@ static int ak_name_gen(char *name, uint8_t index)
+	return NULL;
+}
+
+static int fp_bond_settings_save(struct fp_bond *bond, struct fp_bond *bond_rollback_copy)
```
I doubt if this function should handle the rollback. Maybe we could perform the assignment explicitly in caller to keep single responsibility here (this function unfortunately cannot perform complete rollback on its own)?

alstrzebonski replied:
I don't like this idea. We would have more boilerplate code and risk that someone will eventually not perform rollback.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +735,250 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+	return 0;
+}
+
+static int ak_id_get(uint8_t *id, const struct fp_account_key *account_key)
```
Maybe we could return `id` through return value (`BOND_AK_ID_UNKNOWN` could denote invalid ID)?

alstrzebonski replied:
It would be confusing as BOND_AK_ID_UNKNOWN (or FP_BOND_INFO_AK_ID_UNKNOWN now) is reserved for bond entries that don't have AK associated to them yet. I'll leave it as is.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +735,250 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+	bt_request_cb = cb;
+}
+
+int fp_storage_ak_bond_conn_create(const void *conn_ctx, const bt_addr_le_t *addr,
```
`fp_storage_ak_bond_conn_` - >` fp_storage_ak_bond_` (to simplify API and make it a bit more consistent)

alstrzebonski replied:
We've discussed already that we will use conn infix for API that takes conn_ctx argument. I won't rename it again and again.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +735,250 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+	bond->bonded = true;
+
+	for (size_t i = 0; i < ARRAY_SIZE(fp_bonds); i++) {
+		bond_duplicate = bonded_duplicate_get(conn_ctx, addr);
```
Maybe we could also make a common `bond_duplicate_remove(fp_bond *)` function. Passing `NULL` could result in filtering out all of the duplicates for each bond entry.

alstrzebonski replied:
We pass conn here. I don't understand your idea.

MarekPieta replied:
We get duplicates only to remove them. That's why I suggested to simplify API and just provide a functionality that allows to remove duplicates for a given `fp_bond` entry

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +735,250 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+		return;
+	}
+
+	bond_update_addr(conn_ctx, new_addr);
```
I wonder if we should actually separate `bond_update_addr` function.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +735,250 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+	int err;
+
+	LOG_DBG("Bluetooth bond deleted");
+	bond = bonded_by_addr_get(addr);
```
`addr_to_fp_bond_info` ?

alstrzebonski replied:
`fp_bond_info_by_addr_bonded_get` we only get bonded ones

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -488,13 +1076,45 @@ static int fp_storage_ak_reset(void)
+				 * bt_request_cb->bond_remove on already unpaired bond.
+				 */
+				LOG_WRN("Failed to remove Bluetooth bond because it is already "
+					"unpaired. This might happen in some edge");
```
`in some edge cases`

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -488,13 +1076,45 @@ static int fp_storage_ak_reset(void)
+				continue;
+			}
+
+			err = bt_request_cb->bond_remove(&fp_bonds[index].nv.addr);
```
Consider reusing `bond_remove_completely`/`bond_remove`

alstrzebonski replied:
I'd leave it as is. We clear the settings and RAM explicitly during reset anyway. The callback may not appear if the Fast Pair is disabled so we can't wait for it.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak_bond.h:
```diff
@@ -0,0 +1,142 @@
+ * @param[in] conn_ctx Connection context associated with the Bluetooth bond.
+ * @param[in] addr The Bluetooth address of the bond.
+ */
+void fp_storage_ak_bond_conn_confirm(const void *conn_ctx, const bt_addr_le_t *addr);
```
Maybe we could return error if `conn_ctx` is not valid? Also in other cases. That may simplify spotting problems earlier later on.

MarekPieta replied:
Alternatively we can assert `conn_ctx` validity internally

alstrzebonski replied:
What do you mean by valid? `conn_ctx != NULL` or some interaction with Bluetooth?

MarekPieta replied:
Either NULL or context that was not related to a previously created bond

MarekPieta replied:
Discussed offline. For now we won't do extra validation.

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_auth.c:
```diff
@@ -68,15 +69,25 @@ static void clear_conn_context(const struct bt_conn *conn)
-	return err;
+	if (IS_ENABLED(CONFIG_BT_FAST_PAIR_BOND_MANAGER)) {
+		/* Saving potential bond. */
+		err = fp_keys_bond_save(conn);
```
I'd rather not overcomplicate it. The `fp_keys_bt_auth_progress` is already overloaded and hard to track.

MarekPieta replied:
Discussed offline. I still think `fp_keys_bt_auth_progress` would improve responsibility split among `fp_keys` and `fp_auth`, but I would not insist

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_keys.c:
```diff
@@ -533,5 +534,37 @@ static int fp_keys_uninit(void)
+		account_key = NULL;
+	} else {
+		/* Subsequent pairing - the Account Key is already known. */
+		account_key = (const struct fp_account_key *)proc->aes_key;
```
Then we would have to cast it to be able to use `fp_crypto_aes128_ecb_encrypt` and similar crypto functions. Either way we will need to cast something. I'd keep it as is to have casting in only one place. Also having `aes_key` type as `fp_account_key` may be confusing. In Initial Pairing procedure it's not an account key.

MarekPieta replied:
I am just a bit worried that this cast could go wrong if we append an additional data field to `fp_account_key` structure at some point. That's why I suggested the change.

Alternatively, maybe we could create the struct object here locally and then pass it to function to avoid the risky cast?

MarekPieta replied:
Using a structure here could also help us if we opt for using opaque key identifiers instead of byte arrays here (PSA API)

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/Kconfig.fp_storage:
```diff
@@ -8,6 +8,7 @@ config BT_FAST_PAIR_STORAGE
 	bool
 	default y
 	select SETTINGS
+	select BT_FAST_PAIR_STORAGE_AK
```
We can't do that because of dependency loop:
![image](https://github.com/user-attachments/assets/7b479253-18ea-4bf1-af6c-c19b0097e438)

alstrzebonski replied:
The Kconfig approach will be aligned in different PR by @kapi-no

MarekPieta replied:
Ok. I see that Kconfig options per use-case may help here

alstrzebonski commented:
Adressed review comments. Please rereview.

alstrzebonski commented:
Pushed small fixes.

kapi-no commented on subsys/bluetooth/services/fast_pair/Kconfig.fast_pair:
```diff
@@ -184,6 +186,36 @@ config BT_FAST_PAIR_STORAGE_INTEGRATION_INIT_PRIORITY
 
 endif # BT_FAST_PAIR_STORAGE_INTEGRATION
 
+config BT_FAST_PAIR_BOND_MANAGER
```
Please add it as a `select` or `imply` option to the use case layer for the input device and mouse use case.

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_keys.c:
```diff
@@ -533,5 +534,37 @@ static int fp_keys_uninit(void)
+	__ASSERT_NO_MSG(bt_fast_pair_is_ready());
+
+	struct fp_procedure *proc = &fp_procedures[bt_conn_index(conn)];
+	const struct fp_account_key *account_key;
```
the build system has issues detecting it as uninitialized variable, you can initialize it to NULL just to suppress this warning

kapi-no commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
Looking good. I still think that moving the `fp_bond_info` part to a dedicated source file would be another improvement with a very thin layer of interaction between the AK storage:

```
fp_storage_ak_bond_priv_load_bond_info(...);

fp_storage_ak_bond_priv_ak_save_notify(...);

fp_storage_ak_bond_priv_ak_overwrite_notify(...);

fp_storage_ak_bond_priv_ak_id_mapping_cb_register(...);
```

This private API clearly shows the interdependencies between these two storage entries and gets rid of a lot `IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND)` guards in the code.

Perhaps, we can create a JIRA ticket to work on this improvement in the future.

alstrzebonski replied:
We can think about it. Please don't resolve this comment. I've created task for it: https://nordicsemi.atlassian.net/browse/NCSDK-28604. @MarekPieta let me know what do you think.

MarekPieta replied:
IIRC, the API between `fp_bond_info` and `AK storage` was quite wide and complex (that was my main concern here). If we manage to simplify it (not sure if possible), we could move the implementation to a separate source file/

kapi-no commented:
Please also to create documentation for this change (changelog), align service documentation and possibly integration guide.

alstrzebonski commented:
Rebased on main.

alstrzebonski commented:
Addressed review comments

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include_priv/fp_storage_ak_priv.h:
```diff
@@ -46,6 +46,30 @@ extern "C" {
+#define SETTINGS_BOND_INFO_NAME_MAX_SIZE \
+	(sizeof(SETTINGS_BOND_INFO_FULL_PREFIX) + SETTINGS_BOND_INFO_NAME_MAX_SUFFIX_LEN)
+
+#if CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND && defined(CONFIG_BT_MAX_PAIRED) && \
```
Can we rely only on the first check? (I think both `CONFIG_BT_MAX_PAIRED` and `CONFIG_BT_MAX_CONN` would always be defined)

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_auth.c:
```diff
@@ -212,6 +223,18 @@ static void pairing_complete(struct bt_conn *conn, bool bonded)
+				return;
+			}
+
+			fp_storage_ak_bond_conn_confirm(conn, conn_info.le.dst);
```
I think splitting responsibilities related to `fp_bond` handling in fast_pair service may make it more error prone, but I would not insist here too

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -383,6 +706,28 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+			return -EINVAL;
+		}
+
+		__ASSERT(bond->bonded, "Account Key cannot be written by unbonded connection.");
```
`fp_bond writing an Account Key must be already bonded`? Keep in mind that Account Key is written also by Locator Tags that do not bond at all

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +758,211 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+
+	bond->bonded = true;
+
+	FP_BONDS_FOREACH(bond) {
```
Do we actually need `FP_BONDS_FOREACH` here? Can we expect multiple duplicated bond entries in some cases?

alstrzebonski replied:
You're right we can get only one duplicated bond here.

MarekPieta replied:
Discussed offline. Break was added to handle it properly (initially overlooked it)

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak_bond.h:
```diff
@@ -0,0 +1,131 @@
+ *
+ * @param[in] conn_ctx Connection context associated with the Bluetooth bond.
+ */
+void fp_storage_ak_bond_conn_finalize(const void *conn_ctx);
```
The name suggest that procedure is finalized on disconnection. Maybe we can call it e.g. `fp_storage_ak_bond_conn_disonnected`/`fp_storage_ak_bond_conn_drop`?

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -488,12 +1070,49 @@ static int fp_storage_ak_reset(void)
 	}
 
+	if (IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND)) {
+		FP_BONDS_FOREACH(bond) {
```
Could we reuse `bond_remove` function? Maybe we could even always accept `-ESRCH` returned by `bt_request_cb->bond_remove` to simplify implementation on our end

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include_priv/fp_storage_ak_priv.h:
```diff
@@ -46,6 +46,30 @@ extern "C" {
+/** Length of the Fast Pair bonds arrays. */
+#define FP_BONDS_ARRAY_LEN	(0)
+#endif
+BUILD_ASSERT(FP_BONDS_ARRAY_LEN <= 100);
```
Can we add a comment that it needs to fit on 2 digits?

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -113,6 +166,10 @@ static int fp_settings_set(const char *name, size_t len, settings_read_cb read_c
+	} else if (IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND) &&
+		   !strncmp(name, SETTINGS_BOND_INFO_NAME_PREFIX,
+		   sizeof(SETTINGS_BOND_INFO_NAME_PREFIX) - 1)) {
+		err = fp_settings_load_bond(name, len, read_cb, cb_arg);
```
I will create follow-up ticket for it.

alstrzebonski replied:
https://nordicsemi.atlassian.net/browse/NCSDK-28734

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_bond_manager.c:
```diff
@@ -0,0 +1,160 @@
+
+static int bond_remove(const bt_addr_le_t *addr)
+{
+	__ASSERT_NO_MSG(is_enabled);
```
Would fail if Fast Pair factory reset is called while Fast Pair subsystem is disabled.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,53 +325,204 @@ static int validate_ak_order(void)
+
+	if (bt_stack_bond_remove && bond->bonded) {
+		/* Bond entry will be deleted in the fp_storage_ak_bond_delete function called
+		 * after bond removal from Bluetooth stack.
```
Discussed offline, the `bond_deleted` callback will be be called by Bluetooth stack and we will receive it. Still, we can delete `fp_bond` right after `bt_request_cb->bond_remove` if the bond still exists (to keep common codebase).

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +758,211 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+	bond->bonded = true;
+
+	FP_BONDS_FOREACH(bond) {
+		bond_duplicate = fp_bond_info_bonded_duplicate_get(conn_ctx, addr);
```
I'd past it all in code to not add another remove functions to not cause confusion.

MarekPieta replied:
Looks good already

NordicBuilder commented:
## CI Information

<sup>_To view the history of this post, clich the 'edited' button above_</sup>
Build number: 16
### Inputs:
#### Sources:
sdk-nrf: PR head: [f4de6fbda4c1151b804f054a5cf3fc3662cd059e](https://github.com/nrfconnect/sdk-nrf/pull/15146/commits/f4de6fbda4c1151b804f054a5cf3fc3662cd059e)
<details><summary> more details</summary>

sdk-nrf:
> PR head: [f4de6fbda4c1151b804f054a5cf3fc3662cd059e](https://github.com/nrfconnect/sdk-nrf/pull/15146/commits/f4de6fbda4c1151b804f054a5cf3fc3662cd059e)
> merge base: [b8bb4cc44258779149bcfe98cb3661c90a7aaa4c](https://github.com/nrfconnect/sdk-nrf/commit/b8bb4cc44258779149bcfe98cb3661c90a7aaa4c)
> target head (main): [47e5835fe1e3832b9898d0263a0ecfafdaf8d3c8](https://github.com/nrfconnect/sdk-nrf/commit/47e5835fe1e3832b9898d0263a0ecfafdaf8d3c8)> [Diff](https://github.com/nrfconnect/sdk-nrf/compare/b8bb4cc44258779149bcfe98cb3661c90a7aaa4c..f4de6fbda4c1151b804f054a5cf3fc3662cd059e)
</details>

#### Github labels
| Enabled | Name | Description |
| -- | -- | -- |
|| ci-disabled | _Disable the ci execution_ | 
|| ci-all-test | _Run all of ci, no test spec filtering will be done_ | 
|| ci-force-downstream | _Force execution of downstream even if twister fails_ | 
|| ci-run-twister | _Force run twister_ | 
|| ci-run-zephyr-twister | _Force run zephyr twister_ | 

<details><summary> List of changed files detected by CI (16)</summary>

```
subsys
│  ├── bluetooth
│  │  ├── services
│  │  │  ├── fast_pair
│  │  │  │  ├── CMakeLists.txt
│  │  │  │  ├── Kconfig.fast_pair
│  │  │  │  ├── fp_auth.c
│  │  │  │  ├── fp_bond_manager.c
│  │  │  │  ├── fp_keys.c
│  │  │  │  ├── fp_storage
│  │  │  │  │  ├── Kconfig.fp_storage
│  │  │  │  │  ├── fp_storage_ak.c
│  │  │  │  │  ├── fp_storage_ak_minimal.c
│  │  │  │  │  ├── include
│  │  │  │  │  │  ├── fp_storage_ak.h
│  │  │  │  │  │  │ fp_storage_ak_bond.h
│  │  │  │  │  ├── include_priv
│  │  │  │  │  │  │ fp_storage_ak_priv.h
│  │  │  │  ├── include
│  │  │  │  │  ├── fp_auth.h
│  │  │  │  │  │ fp_keys.h
│  │  │  │  ├── use_case
│  │  │  │  │  │ Kconfig
tests
│  ├── subsys
│  │  ├── bluetooth
│  │  │  ├── fast_pair
│  │  │  │  ├── storage
│  │  │  │  │  ├── account_key_storage
│  │  │  │  │  │  ├── src
│  │  │  │  │  │  │  │ main.c
│  │  │  │  │  ├── common
│  │  │  │  │  │  ├── src
│  │  │  │  │  │  │  │ common_utils.c
```
</details>

### Outputs:

#### Toolchain
Version: 
Build docker image: 

#### Test Spec & Results: :white_check_mark: Success; :x: Failure; :orange_circle: Queued; :yellow_circle: Progress; :white_medium_square: Skipped
- :orange_circle: Toolchain
- :orange_circle: Build twister
- :white_medium_square: Integration tests
   - :orange_circle: desktop52_verification
   - :orange_circle: test_ble_commit
   - :orange_circle: test-fw-nrfconnect-ble_samples
   - :orange_circle: test-sdk-find-my

<details><summary>Disable integration tests</summary>

- 
   - doc-internal
   - latest/ble/nrf/test_ble_commit
   - test-fw-nrfconnect-apps
   - test-fw-nrfconnect-ble_mesh
   - test-fw-nrfconnect-boot
   - test-fw-nrfconnect-chip
   - test-fw-nrfconnect-fem
   - test-fw-nrfconnect-nfc
   - test-fw-nrfconnect-nrf-iot_cloud
   - test-fw-nrfconnect-nrf-iot_libmodem-nrf
   - test-fw-nrfconnect-nrf-iot_lwm2m
   - test-fw-nrfconnect-nrf-iot_mosh
   - test-fw-nrfconnect-nrf-iot_nrf_provisioning
   - test-fw-nrfconnect-nrf-iot_positioning
   - test-fw-nrfconnect-nrf-iot_samples
   - test-fw-nrfconnect-nrf-iot_serial_lte_modem
   - test-fw-nrfconnect-nrf-iot_thingy91
   - test-fw-nrfconnect-nrf-iot_zephyr_lwm2m
   - test-fw-nrfconnect-nrf_crypto
   - test-fw-nrfconnect-proprietary_esb
   - test-fw-nrfconnect-ps
   - test-fw-nrfconnect-rpc
   - test-fw-nrfconnect-rs
   - test-fw-nrfconnect-tfm
   - test-fw-nrfconnect-thread
   - test-fw-nrfconnect-zigbee
   - test-low-level
   - test-sdk-audio
   - test-sdk-dfu
   - test-sdk-mcuboot
   - test-sdk-pmic-samples
   - test-sdk-sidewalk
   - test-sdk-wifi
</details>



*Note: This message is automatically posted and updated by the CI*

alstrzebonski commented:
Addressed review comments. Please rereview.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,53 +325,203 @@ static int validate_ak_order(void)
-			id = ACCOUNT_KEY_METADATA_FIELD_GET(account_key_metadata[i], ID);
-			if (id != 0) {
-				return -EINVAL;
+		/* The bond might have already been removed by fp_storage_ak_bond_delete. */
```
The module that handles Bluetooth bond request callbacks calls fp_storage_ak_bond_delete to remove fp_bond_info when a Bluetooth bond is deleted for any reason. Because of this, the fp_bond_info might have already been cleared at this point.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,53 +325,204 @@ static int validate_ak_order(void)
 	}
+}
+
+static void fp_settings_validate_fp_bond_info(void)
```
Discussed offline. We will forward error codes here too

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +758,211 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+	struct fp_bond_info bond_rollback_copy;
+
+	bond = fp_bond_info_by_conn_get(conn_ctx);
+	if (!bond) {
```
I see that in some functions we need to silently skip case when `fp_bond` record was not found (`fp_bond_manager` does not tracks ongoing procedures on its own so we may get request related to non Fast Pair BLE connection)

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_bond_manager.c:
```diff
@@ -0,0 +1,164 @@
+/* Initialize the module on boot to be able to remove Bluetooth bonds during Fast Pair storage
+ * factory reset even if the Fast Pair is not enabled.
+ */
+SYS_INIT(bond_manager_init, APPLICATION, CONFIG_APPLICATION_INIT_PRIORITY);
```
I think we should not fully initialize the module during `SYS_INIT`. Please only register the `bt_request_cb` (as it's necessary to allow removing Bluetooth bonds while Fast Pair is disabled)

alstrzebonski commented:
Addressed review comments. Please rereview.

MarekPieta commented on tests/subsys/bluetooth/fast_pair/storage/account_key_storage/src/main.c:
```diff
@@ -93,7 +93,7 @@ ZTEST(suite_fast_pair_storage_common, test_one_key)
 
 	cu_generate_account_key(seed, &account_key);
-	err = fp_storage_ak_save(&account_key);
+	err = fp_storage_ak_save(&account_key, NULL);
```
Consider adding e.g. a build assert to clearly indicate that AK storage tests do not support Fast Pair bond management (similar limitation may apply to the factory reset test)

alstrzebonski commented:
Fixed compliance and rebased on main.

MarekPieta commented on subsys/bluetooth/services/fast_pair/Kconfig.fast_pair:
```diff
@@ -184,6 +186,36 @@ config BT_FAST_PAIR_STORAGE_INTEGRATION_INIT_PRIORITY
+	range 0 9
+	default 3
+	help
+	  Fast Pair bond manager module initialization priority.
```
Do we still need to use lower value for initialization priority here? The callbacks registration that enforced initialization order is now moved to `SYS_INIT`

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_bond_manager.c:
```diff
@@ -0,0 +1,174 @@
+	return 0;
+}
+
+BUILD_ASSERT(CONFIG_BT_FAST_PAIR_BOND_MANAGER_INIT_PRIORITY <
```
No longer true I think

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_auth.c:
```diff
@@ -68,15 +69,25 @@ static void clear_conn_context(const struct bt_conn *conn)
-	return err;
+	if (IS_ENABLED(CONFIG_BT_FAST_PAIR_BOND_MANAGER)) {
+		/* Saving potential bond. */
+		err = fp_keys_bond_save(conn);
```
I also wonder if we shouldn't call `fp_storage_ak_bond_conn_finalize` also when pairing fails (are we sure that failed pairing would always result in disconnection?)

alstrzebonski replied:
As for 2nd comment. We should be fine to call it on disconnection. The approach is to clear bonds on disconnection to have just one synchronization point.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +755,222 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+	return 0;
+}
+
+void fp_storage_ak_bond_conn_confirm(const void *conn_ctx, const bt_addr_le_t *addr)
```
Consider validating that `addr` from the function argument matches address stored for the `bond_confirmed`. Maybe we could even skip passing `addr` as an argument here (assuming that we can rely on validity of address stored for `bond_confirmed`)

alstrzebonski replied:
I'll add assert here.

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -361,6 +677,10 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
 	char name[SETTINGS_AK_NAME_MAX_SIZE];
 	int err;
 
+	struct fp_bond_info *bond;
+	struct fp_bond_info bond_rollback_copy;
```
Only bond_rollback_copy can be scoped locally.

MarekPieta replied:
You are right. The `conn_ctx` assignment is done outside of this block

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +755,222 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+		struct fp_bond_info *bond;
+		int err;
+
+		/* In some edge cases, while enabling and validating bonds there might be more than
```
Those loops call bt_unpair which calls this function callback so we can land here when handling bond duplicates during Fast Pair enable operation. We can't firstly remove the bond duplicates from the Fast Pair storage because of power down risk.

MarekPieta replied:
You are right. Let's keep it as is.

alstrzebonski commented:
Addressed review comments and rebased on main.

MarekPieta commented on tests/subsys/bluetooth/fast_pair/storage/account_key_storage/src/main.c:
```diff
@@ -18,6 +18,7 @@
 
 #define ACCOUNT_KEY_MAX_CNT	CONFIG_BT_FAST_PAIR_STORAGE_ACCOUNT_KEY_MAX
 
+BUILD_ASSERT(!IS_ENABLED(CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND));
```
Consider commenting that it's because the AK bond storage feature is not yet supported by the unit tests (to clarify why the assert was introduced here)

alstrzebonski replied:
Done.

MarekPieta commented on subsys/bluetooth/services/fast_pair/Kconfig.fast_pair:
```diff
@@ -184,6 +186,36 @@ config BT_FAST_PAIR_STORAGE_INTEGRATION_INIT_PRIORITY
+config BT_FAST_PAIR_BOND_MANAGER
+	bool "Fast Pair bond management functionality [EXPERIMENTAL]"
+	default y
+	depends on !BT_FAST_PAIR_FMDN
```
Shouldn't we introduce `BT_FAST_PAIR_USE_CASE_UNSUPPORTED_BOND_MANAGER` instead of depending on `!BT_FAST_PAIR_FMDN`? That could improve consistency with approach taken for other Kconfigs. @kapi-no what do you think?

MarekPieta replied:
You are right. Seems that I overlooked it.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -405,9 +747,223 @@ int fp_storage_ak_save(const struct fp_account_key *account_key)
+		 */
+		bond = fp_bond_info_by_addr_bonded_get(addr);
+		if (!bond) {
+			return;
```
Consider using `break` here to avoid this early return

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -435,6 +996,25 @@ static int fp_storage_ak_init(void)
+			 * Fast Pair Procedure disconnected, there may be a conn_ctx not set to
+			 * NULL. All conn_ctx should be set to NULL at this point.
+			 */
+			bond->conn_ctx = NULL;
```
Shouldn't we somehow ensure that this incomplete bond record will be removed? Keep in mind that the same peer might have bonded without following Fast Pair flow after Fast Pair was disabled (unless I missed something, it would not be properly removed if the interrupted Fast Pair Procedure was related to subsequent pairing)

alstrzebonski replied:
I'm afraid it's really hard to do something about it. It's not really severe. Worst that can happen is that on AK removal we remove some bond that started subsequent pairing procedure when Fast Pair was enabled but finished the procedure when Fast Pair was disabled.

alstrzebonski replied:
I can create task to improve it in the future.

MarekPieta replied:
Discussed offline. The edge case will be documented and @alstrzebonski will open a follow-up ticket to fix it in the future (proper fix may require tracking Fast Pair BLE bonds also while Fast Pair is disabled - initially we wanted to avoid that but it may be worth reconsidering this approach as we may be unable to prevent some of the edge cases without it).

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak_minimal.c:
```diff
@@ -141,11 +141,13 @@ int fp_storage_ak_find(struct fp_account_key *account_key,
 	int err;
 	struct account_key_data data;
 
+	__ASSERT_NO_MSG(!conn_ctx);
```
`ARG_UNUSED` -> we pass the conn object unconditionally here, so the assertion would actually fail

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak_bond.h:
```diff
@@ -0,0 +1,143 @@
+ */
+void fp_storage_ak_bond_conn_finalize(const void *conn_ctx);
+
+/** Update the address of the Bluetooth bond associated with the Account Key.
```
`update` -> `resolve` ? To make it more straightforward

alstrzebonski replied:
This function is not really resolving anything. It updates the address with the address that was already resolved earlier.

MarekPieta replied:
Ok, let's keep the naming from perspective of the `fp_storage`. Thanks for adding the comment

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak_bond.h:
```diff
@@ -0,0 +1,143 @@
+ *  persist in the memory after this function exits, as it is used directly without the copy
+ *  operation. It is possible to register only one instance of callbacks with this API.
+ *
+ *  This function must be called before initialiazing the Fast Pair storage with the
+ *  @ref fp_storage_init API if the CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND Kconfig option is enabled.
+ *  This function must be called before performing the Fast Pair factory reset if the
+ *  CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND Kconfig option is enabled.
```
`This function should be registered early to ensure that the callbacks are available both during Fast Pair factory reset and on Fast Pair storage initialization`? I think there is no need to mention `CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND` as whole header is used only if `CONFIG_BT_FAST_PAIR_STORAGE_AK_BOND` is enabled

alstrzebonski replied:
Changed to
```
 *  This function must be called before initialiazing the Fast Pair storage with the
 *  @ref fp_storage_init API and also before performing the Fast Pair factory reset to ensure that
 *  the callbacks are available during those operations.
```

MarekPieta commented on subsys/bluetooth/services/fast_pair/use_case/Kconfig:
```diff
@@ -75,6 +76,10 @@ config BT_FAST_PAIR_USE_CASE_MOUSE
 endchoice
 
 # Collection of unsupported Fast Pair features that can be selected by the use case Kconfig.
+config BT_FAST_PAIR_USE_CASE_UNSUPPORTED_BOND_MANAGER
```
This seems to belong to the previous commit (in general the first 2 commits seem a bit mixed up now)

alstrzebonski replied:
I've merged those first two commits. They're interdependent anyway.

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_bond_manager.c:
```diff
@@ -0,0 +1,166 @@
+
+	int err;
+
+	is_enabled = false;
```
It'd be inconsistent with other fast_pair modules and also would not allow for enabling Fast Pair in case of uninit failure. The module would also think that it's enabled even though it's in the middle of disabling.

alstrzebonski commented on subsys/bluetooth/services/fast_pair/fp_storage/fp_storage_ak.c:
```diff
@@ -224,56 +325,222 @@ static int validate_ak_order(void)
+		if (err) {
+			if (err == -ESRCH) {
+				LOG_DBG("Bond not found in the Bluetooth stack. "
+					"Continuing the removal operation");
```
That's what the log is for. If I'd leave only the `"Bond not found in the Bluetooth stack. "` message it could give the false impression that the removal operation is aborted if someone calls `bond_remove` with `bt_stack_bond_remove` set to true if the bond is not present in the BT stack.

MarekPieta replied:
Ok, we can keep it as is

alstrzebonski commented:
Addressed review comments. Please rereview.

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak_bond.h:
```diff
@@ -0,0 +1,143 @@
+ *  established from the active connection.
+ *
+ *  The module links the Account Key information with the Bluetooth bond and stores
+ *  it in the NVM memory. The association and the Bluetooth bond will be deleted if the bond is not
```
`in the NVM.`

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak_bond.h:
```diff
@@ -0,0 +1,143 @@
+
+/** Delete the Account Key association with the Bluetooth bond.
+ *
+ *  If the @ref fp_storage_ak_bond_conn_create API has not been called for this conn earlier,
```
There is no `conn` in function argument list.

MarekPieta replied:
Consider explicitly mentioning that the function needs to be called when a Bluetooth bond is removed to let `fp_storage` perform the needed cleanup of AK and BLE bond associations.

MarekPieta replied:
You can also state that function is no-op if there is no association with AK

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak_bond.h:
```diff
@@ -0,0 +1,143 @@
+
+/** Indicate that the potential Bluetooth bond has been confirmed.
+ *
+ *  It confirms the potential Bluetooth bond associated with the Account Key. The association should
```
`with an Account Key`?

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include/fp_storage_ak_bond.h:
```diff
@@ -0,0 +1,143 @@
+ *
+ *  It relates to the Account Key association with the Bluetooth bond that was created
+ *  earlier using the @ref fp_storage_ak_bond_conn_create API. This function should be called when
+ *  the BLE conn disconnects to let the storage subsystem handle incomplete bonds.
```
`handle` -> `clean up`?

MarekPieta commented on subsys/bluetooth/services/fast_pair/fp_storage/include_priv/fp_storage_ak_priv.h:
```diff
@@ -46,6 +46,30 @@ extern "C" {
 #define SETTINGS_AK_NAME_MAX_SIZE \
 	(sizeof(SETTINGS_AK_FULL_PREFIX) + SETTINGS_AK_NAME_MAX_SUFFIX_LEN)
 
+/** Settings key name prefix for bond info. */
```
`for bond` (for consistency with subsequent comments)

alstrzebonski commented:
Addressed review comments and tested the solution in details. The documentation will be added in separate PR.

alstrzebonski commented:
Rebased on main.

MarekPieta reviewed (approved):
Looks good. Thank you for the change!

alstrzebonski commented:
Rebased on main.

alstrzebonski commented:
Rebased on main.
